File: apps/mobile/modules/ryco-native-controls/android/src/main/java/expo/modules/ryconativecontrols/RycoNativeControlsModule.kt
package expo.modules.ryconativecontrols

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class RycoNativeControlsModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("RycoNativeControls")

    Function("getShowcasePairingUrl") {
      appContext.currentActivity?.intent?.getStringExtra("showcasePairingUrl")
    }

    Function("getShowcaseScene") {
      val storedScene = appContext.reactContext
        ?.filesDir
        ?.resolve("t3-showcase-scene")
        ?.takeIf { it.isFile }
        ?.readText()
        ?.trim()
        ?.takeIf { it.isNotEmpty() }
      storedScene ?: appContext.currentActivity?.intent?.getStringExtra("showcaseScene")
    }

    Function("prepareShowcaseCapture") {
      // Android app data is cleared by the host runner before launch.
    }

    Function("markShowcaseReady") { scene: String ->
      appContext.reactContext
        ?.filesDir
        ?.resolve("t3-showcase-ready")
        ?.writeText(scene)
    }

    View(RycoHeaderButtonView::class) {
      Prop("label") { view: RycoHeaderButtonView, label: String ->
        view.setLabel(label)
      }
      Prop("systemImage") { view: RycoHeaderButtonView, systemImage: String ->
        view.setSystemImage(systemImage)
      }

      Events("onTriggered")
    }
  }
}

File: apps/mobile/modules/ryco-native-controls/ios/RycoNativeControlsModule.swift
import ExpoModulesCore

public final class RycoNativeControlsModule: Module {
  public func definition() -> ModuleDefinition {
    Name("RycoNativeControls")

    // Upstream's screenshot/showcase-rig functions (getShowcaseScene,
    // markShowcaseReady, prepareShowcaseCapture, getShowcasePairingUrl) are
    // omitted: the showcase rig is stripped from the Ryco MVP.
    View(RycoHeaderButtonView.self) {
      Prop("label") { (view: RycoHeaderButtonView, label: String) in
        view.setLabel(label)
      }
      Prop("systemImage") { (view: RycoHeaderButtonView, systemImage: String) in
        view.setSystemImage(systemImage)
      }

      Events("onTriggered")
    }
  }
}
